src/service.js

```diff
@@ -425,24 +462,54 @@ function getServiceInternal(holder, context, id, opt_factory,
       obj: null,
       promise: null,
       resolve: null,
+      ctor: null,
     };
   }
 
   if (!s.obj) {
-    dev().assert(opt_factory || opt_constructor,
-        'Factory or class not given and service missing %s', id);
-    s.obj = opt_constructor
-        ? new opt_constructor(context)
-        : opt_factory(context);
-    // The service may have been requested already, in which case we have a
-    // pending promise we need to fulfill.
-    if (s.resolve) {
-      s.resolve(s.obj);
+    if (s.ctor) {
+      const ctor = s.ctor;
+      s.obj = new ctor(context);
+    } else {
+      // TODO(kmh287): Replace opt_constructor param with ctor on service
```

Comment:
Yes. Eventually the opt_constructor argument can be removed entirely, simplifying this.